wavefront/resource_metrics_policy.go

```diff
@@ -0,0 +1,347 @@
+
+		if err := validatePolicySelectorSet(accountIds, userGroupIds, roleIds); err != nil {
+			return nil, err
+		}
+
+		newRule := wavefront.PolicyRuleRequest{
+			AccountIds:   accountIds,
+			UserGroupIds: userGroupIds,
+			RoleIds:      roleIds,
+			Name:         rule[nameKey].(string),
+			Tags:         parsePolicyTagsArr(rule[tagsKey]),
+			Description:  rule[descriptionKey].(string),
+			Prefixes:     parseStrArr(rule[prefixesKey]),
+			TagsAnded:    rule[tagsAndedKey].(bool),
+			AccessType:   rule[accessTypeKey].(string),
+		}
+
+		rules = append(rules, newRule)
+	}
+	return rules, nil
+}
+
+func validatePolicySelectorSet(accountIds, userGroupIds, roleIds []string) error {
+	if len(accountIds)+len(userGroupIds)+len(roleIds) < 1 {
+		return errors.New("policy_rule must have at least one associated account, user group, or role")
+	}
+	return nil
+}
+
+func parsePolicyTagsArr(raw interface{}) []wavefront.PolicyTag {
+	var arr []wavefront.PolicyTag
+	if raw != nil && len(raw.([]interface{})) > 0 {
+		for _, v := range raw.([]interface{}) {
+			kv := v.(map[string]interface{})
+			key := kv[policyTagKey].(string)
+			value := kv[policyTagValue].(string)
+			arr = append(arr, wavefront.PolicyTag{
+				Key:   key,
+				Value: value,
+			})
+		}
+	}
+	return arr
+}
+
+// resourceMetricsPolicyDelete reverts metrics policy to default predefined policy rule allowing access to all metrics for everyone
+func resourceMetricsPolicyDelete(d *schema.ResourceData, meta interface{}) error {
+	// needed to lookup default 'everyone' group assignment
+	groups := meta.(*wavefrontClient).client.UserGroups()
+	groupResults, err := groups.Find(
+		[]*wavefront.SearchCondition{
+			{
+				Key:            "name",
+				Value:          "Everyone",
+				MatchingMethod: "EXACT",
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("error reading Default UserGroup 'Everyone' in Wavefront, %s", err)
+	}
+
+	if len(groupResults) != 1 {
+		return fmt.Errorf("error finding default UserGroup 'Everyone' in Wavefront")
+	}
+
+	defaultGroup := groupResults[0]
+
+	metrics := meta.(*wavefrontClient).client.MetricsPolicyAPI()
+
+	defaultPolicyRules := &wavefront.UpdateMetricsPolicyRequest{
+		PolicyRules: []wavefront.PolicyRuleRequest{{
+			UserGroupIds: []string{*defaultGroup.ID},
+			Name:         "Allow All Metrics",
+			Description:  "Predefined policy rule. Allows access to all metrics (timeseries, histograms, and counters) for all accounts. If this rule is removed, all accounts can access all metrics if there are no matching blocking rules.",
+			Prefixes:     []string{"*"},
+			TagsAnded:    false,
+			AccessType:   "ALLOW",
+		}},
+	}
+	_, err = metrics.Update(defaultPolicyRules)
+	if err != nil {
+		return fmt.Errorf("error deleting custom metrics policy: %d", err)
+	}
+	d.SetId("")
+	return nil
+}
+
+// parseStrArr parses a raw interface from d *schema.ResourceData that contains an array of strings
+func parseStrArr(raw interface{}) []string {
```

Comment:
FYI, there is a getStringSlice() function in utils.go that does essentially the same thing as this.  Except that getStringSlice() is expecting the list to be in a *schema.Set. I don't know why you don't have to deal with schema.Set in your code.  But, I am wondering if you could put this function in utils.go and then have getStringSlice() call into it?